chaosgcp/apphub/actions.py

```diff
@@ -0,0 +1,315 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from typing import Dict, Any, List, Generator, Optional
+
+from google.cloud import compute_v1, apphub_v1
+from google.cloud import resourcemanager_v3
+
+from chaoslib.types import Configuration, Secrets
+from chaosgcp.lb.actions import (
+    inject_traffic_faults,
+    remove_fault_injection_traffic_policy,
+)
+from chaosgcp import get_context, load_credentials
+from chaoslib.exceptions import ActivityFailed
+
+
+__all__ = [
+    "inject_fault_if_url_map_exists_app_hub",
+    "remove_fault_if_url_map_exists_app_hub",
+]
+logger = logging.getLogger("chaostoolkit")
+
+
+def apphub_app_list_services(
+    host_project_id: str,
+    project_id: str,
+    location: str,
+    application_id: str,
+    configuration: Configuration = None,
+    secrets: Secrets = None,
+) -> List[str]:
+    """Retrieves a list of service URIs associated with an AppHub application.
+
+    This function queries the AppHub API to list all services deployed within
+    a specific application and extracts their associated URIs.
+
+    Args:
+        host_project_id (str): The ID of the Host Apphub project.
+        project_id (str): The ID of the Google Cloud project to inject fault into.
+        location (str): The geographic location (e.g., "us-central1") of the application.
+        application_id (str): The unique identifier for the AppHub application.
+
+    Returns:
+        list: A list of strings, where each string is the URI of a service associated
+              with the specified application. The URIs are formatted to contain the
+                      project ID instead of the project number only for project_id."""
+    try:
+        name_app = f"projects/{host_project_id}/locations/{location}/applications/{application_id}"
+        credentials = load_credentials(secrets)
+        client = apphub_v1.AppHubClient()
+        request = apphub_v1.ListServicesRequest(parent=name_app)
+        service_uris = []
+        updated_uri = None
+        logger.info("Getting attached Services of given Application in Apphub")
+        for response in client.list_services(request=request):
+            uri = response.service_reference.uri.replace(
+                resourcemanager_v3.ProjectsClient(credentials=credentials)
+                .get_project(name="projects/{}".format(project_id))
+                .name,
+                "projects/{}".format(project_id),
+            )
+            updated_uri = (
+                uri[uri.find("/projects/") :] if "/projects/" in uri else uri
+            )
+            service_uris.append(updated_uri)
+        logger.debug(f"Listing Services {updated_uri}")
+        return service_uris
+    except Exception as e:
+        raise ActivityFailed(
+            f"Error Listing Services of Apphub Application: {e}"
+        )
+
+
+def get_url_map_essentials(
```

Comment:
this also could be an interesting probe, couldn't it?

Reply:
Yes, Kindly refer the previous comment on probe .